rfcs/20200710-reduction/README.md

```diff
@@ -0,0 +1,159 @@
+  - L-2 normalization
+
+In addition to L-2 the following Normalize agorithms are nice-to-have:
+  - L-1 normalization
+  - L-infinity normalization
+
+The following Reduce and Norm algorithms are supported by the most popular
+frameworks:
+
+| Algotirhm | Formula                      | TensorFlow | PyTorch | OpenVino |
+| ---------:| ----------------------------:| ----------:| -------:| --------:|
+| And       | and(x[:])                    |            |         | x        |
+| Or        | or(x[:])                     |            |         | x        |
+| Min       | min(x[:])                    | x          | x       | x        |
+| Max       | max(x[:])                    | x          | x       | x        |
+| Sum       | sum(x[:])                    | x          | x       | x        |
+| Prod      | mul(x[:])                    | x          | x       | x        |
+| Mean      | sum(x[:]) / n                | x          | x       | x        |
+| Variance  | Mean(pow(x[:] - Mean(x), 2)) | x          | x       |          |
+| Std       | pow(Variance(x), 0.5)        | x          | x       |          |
+| LogSumExp | log(sum(exp(x[:] - Max(x)))) + Max(x) | x | x       |          |
+| Median    |                              |            | x       |          |
+| Mode      |                              |            | x       |          |
+| p-norm    | pow(sum(pow(abs(x[:]), p)), 1/p) | x      | x       |          |
+
+The following algorithms are out of scope due to low priority:
+* And;
+* Or;
+* Median;
+* Mode.
+
+Also there are normalize operations, which variate between frameworks:
+* OpenVino:
+  L2_normalize(x) = x[:] / pow(eps_mode(sum(pow(x[:], 2)), eps), 0.5)
+    = x[:] / pow(eps_mode((pow(norm_l2(x[:]), 2)), eps), 0.5),
+  where eps_mode = {add, max}
+* TensorFlow:
+  L2_normalize(x) = x[:] / pow(max(sum(pow(x[:], 2)), eps), 0.5)
+    = x[:] / pow(max((pow(norm_l2(x[:]), 2)), eps), 0.5)
+* PyTorch:
+  L2_normalize(x) = x[:] / max(pow(sum(pow(x[:], 2)), 0.5), eps)
+    = x[:] / max(norm_l2(x[:]), eps)
+
+# Proposal
+## Option 1: Implement reduction primitive and norm primitive.
+
+### Motivation
+Motivation for this option is to keep reduction primitive API simple, but limit
+flexibility and amount of supported algorithms.
+
+### API
+
+``` cpp
+// dnnl_types.h
+typedef enum {
+    ...
+    dnnl_eltwise_pow,
+    ...
+    dnnl_reduction_max,
+    dnnl_reduction_min,
+    dnnl_reduction_norm_l1,
+    dnnl_reduction_norm_l2,
+    dnnl_reduction_norm_linf,
+    dnnl_reduction_sum,
+    dnnl_reduction_mul,
+    dnnl_reduction_mean,
+    ...
+} dnnl_alg_kind_t;
+
+typedef struct {
+    dnnl_primitive_kind_t primitive_kind;
+    // The kind of reduction algorithm.
+    dnnl_alg_kind_t reduction_alg_kind;
+    // Source memory descriptor.
+    dnnl_memory_desc_t src_desc;
+    // Destination memory descriptor.
+    dnnl_memory_desc_t dst_desc;
+} dnnl_reduction_desc_t;
+
+// dnnl.h
+dnnl_status_t DNNL_API dnnl_reduction_desc_init(
```

Comment:
Just to confirm -- we do not plan to add backward pass for the reduction? Why?

Reply:
I was thinking about reduction as a utility functionality like binary, which can be used to build forward and backward pass of some operation. Do you anticipate that backward pass for reduction itself is important?

Reply:
no, after a second thought I think we don't need it. in the worst case -- binary to the rescue :)